finagle-thrift/src/main/scala/com/twitter/finagle/thrift/MultiplexedService.scala

```diff
@@ -0,0 +1,117 @@
+package com.twitter.finagle.thrift
+
+import com.twitter.finagle.{Service, Thrift}
+import com.twitter.scrooge.TReusableMemoryTransport
+import com.twitter.util.Future
+import java.util.Arrays
+import org.apache.thrift.TApplicationException
+import org.apache.thrift.protocol._
+import org.apache.thrift.transport.TMemoryInputTransport
+import scala.collection.JavaConverters._
+import scala.util.Try
+
+class MultiplexedFinagleService(
+  services: Map[String, Service[Array[Byte], Array[Byte]]],
+  protocolFactory: TProtocolFactory,
+  maxThriftBufferSize: Int = Thrift.Server.maxThriftBufferSize)
+    extends Service[Array[Byte], Array[Byte]] {
+
+  private val serviceMap = services.mapValues(getFunctionMap)
+
+  def getFunctionMap(
+    service: Service[Array[Byte], Array[Byte]]
+  ): collection.Map[String, (TProtocol, Int) => Future[Array[Byte]]] = {
+    Try(scroogeFinagleServiceFunctionMap(service))
+      .orElse(Try(thriftFinagleServiceFunctionMap(service)))
+      .getOrElse(throw new IllegalArgumentException("%s cannot be multiplexed".format(service.getClass.getName)))
+  }
+
+  def scroogeFinagleServiceFunctionMap[T](target: AnyRef) = {
```

Comment:
why do we need to support both this and `thriftFinagleServiceFunctionMap`? (mind leaving a comment in the code as a response?)